libkbfs/folder_block_manager.go

```diff
@@ -320,82 +335,59 @@ func (fbm *folderBlockManager) deleteBlockRefs(ctx context.Context,
+		rmds[0].data.Dir, toDelete.md.data.Dir)
+	if err != nil {
+		fbm.log.CErrorf(ctx, "Error when comparing dirs: %v", err)
+	} else if dirsEqual {
+		// This md is part of the history of the folder,
+		// so we shouldn't delete the blocks.
+		fbm.log.CDebugf(ctx, "Not deleting blocks from revision %d",
+			toDelete.md.Revision)
+		// But, since this MD put seems to have succeeded, we
+		// should archive it.
+		fbm.log.CDebugf(ctx, "Archiving successful MD revision %d",
+			rmds[0].Revision)
+		// Don't block on archiving the MD, because that could
+		// lead to deadlock.
+		fbm.archiveUnrefBlocksNoWait(rmds[0])
 		return nil
 	}
 
-	toDeleteAgain := make(map[*RootMetadata][]BlockPointer)
-	for md, ptrs := range toDelete {
-		fbm.log.CDebugf(ctx, "Checking deleted blocks for revision %d",
-			md.Revision)
-		// Make sure that the MD didn't actually become
-		// part of the folder history.  (This could happen
-		// if the Sync was canceled while the MD put was
-		// outstanding.)
-		rmds, err := getMDRange(ctx, fbm.config, fbm.id, md.BID,
-			md.Revision, md.Revision, md.MergedStatus())
-		if err != nil || len(rmds) == 0 {
-			toDeleteAgain[md] = ptrs
-			continue
-		}
-		dirsEqual, err := CodecEqual(fbm.config.Codec(),
-			rmds[0].data.Dir, md.data.Dir)
-		if err != nil {
-			fbm.log.CErrorf(ctx, "Error when comparing dirs: %v", err)
-		} else if dirsEqual {
-			// This md is part of the history of the folder,
-			// so we shouldn't delete the blocks.
-			fbm.log.CDebugf(ctx, "Not deleting blocks from revision %d",
-				md.Revision)
-			// But, since this MD put seems to have succeeded, we
-			// should archive it.
-			fbm.log.CDebugf(ctx, "Archiving successful MD revision %d",
-				rmds[0].Revision)
-			// Don't block on archiving the MD, because that could
-			// lead to deadlock.
-			fbm.archiveUnrefBlocksNoWait(rmds[0])
-			continue
-		}
-
-		// Otherwise something else has been written over
-		// this MD, so get rid of the blocks.
-		fbm.log.CDebugf(ctx, "Cleaning up blocks for failed revision %d",
-			md.Revision)
+	// Otherwise something else has been written over
+	// this MD, so get rid of the blocks.
+	fbm.log.CDebugf(ctx, "Cleaning up blocks for failed revision %d",
+		toDelete.md.Revision)
 
-		_, err = fbm.deleteBlockRefs(ctx, md, ptrs)
-		// Ignore permanent errors
-		_, isPermErr := err.(BServerError)
-		_, isNonceNonExistentErr := err.(BServerErrorNonceNonExistent)
-		if err != nil {
-			fbm.log.CWarningf(ctx, "Couldn't delete some ref in batch %v: %v", ptrs, err)
-			if !isPermErr && !isNonceNonExistentErr {
-				toDeleteAgain[md] = ptrs
-			}
+	_, err = fbm.deleteBlockRefs(ctx, toDelete.md, toDelete.blocks)
+	// Ignore permanent errors
+	_, isPermErr := err.(BServerError)
+	_, isNonceNonExistentErr := err.(BServerErrorNonceNonExistent)
+	if err != nil {
+		fbm.log.CWarningf(ctx, "Couldn't delete some ref in batch %v: %v", toDelete.blocks, err)
+		if !isPermErr && !isNonceNonExistentErr {
+			fbm.enqueueBlockToDelete(toDelete)
```

Comment:
Whoa I didn't think that far. Thanks for catching that!

What do you think if we just do a `go fbm.enqueueBlockToDelete(toDelete)`? It would 1) still insert it to the end of queue, allowing more time before it tries again (Maybe this can be useful when a mdserver or bserver is temporarily under high load and not responding); 2) avoid having long recursive calls when the `chan` is full.

Although for 2), it's not very convincing because it's essentially trading that by spawning a lot of goroutines. I don't know which is worse.